plugins/inputs/prometheus_scraper/metrics_type_handler.go

```diff
@@ -0,0 +1,197 @@
+)
+
+const (
+	prometheusMetricTypeKey = "prom_metric_type"
+
+	histogramSummaryCountSuffix = "_count"
+	histogramSummarySumSuffix   = "_sum"
+	histogramBucketSuffix       = "_bucket"
+)
+
+var (
+	histogramSummarySuffixes = []string{histogramSummaryCountSuffix, histogramSummarySumSuffix, histogramBucketSuffix}
+)
+
+// Get the metric name in the TYPE comments for Summary and Histogram
+// e.g # TYPE nginx_ingress_controller_request_duration_seconds histogram
+//     # TYPE nginx_ingress_controller_ingress_upstream_latency_seconds summary
+func normalizeMetricName(name string) string {
+	for _, s := range histogramSummarySuffixes {
+		if strings.HasSuffix(name, s) && name != s {
+			return strings.TrimSuffix(name, s)
+		}
+	}
+	return name
+}
+
+func (pm *PrometheusMetric) isCounter() bool {
+	return pm.metricType == textparse.MetricTypeCounter
+}
+
+func (pm *PrometheusMetric) isGauge() bool {
+	return pm.metricType == textparse.MetricTypeGauge
+}
+
+func (pm *PrometheusMetric) isHistogram() bool {
+	return pm.metricType == textparse.MetricTypeHistogram
+}
+
+func (pm *PrometheusMetric) isSummary() bool {
+	return pm.metricType == textparse.MetricTypeSummary
+}
+
+// Adapter to prometheus scrape.Target
+type metadataCache interface {
+	Metadata(metricName string) (scrape.MetricMetadata, bool)
+}
+
+// adapter to get metadata from scrape.Target
+type mCache struct {
+	t *scrape.Target
+}
+
+func (m *mCache) Metadata(metricName string) (scrape.MetricMetadata, bool) {
+	return m.t.Metadata(metricName)
+}
+
+// Adapter to ScrapeManager to retrieve the cache by job and instance
+type metadataService interface {
+	Get(job, instance string) (metadataCache, error)
+}
+
+type metadataServiceImpl struct {
+	sm ScrapeManager
+}
+
+func (t *metadataServiceImpl) Get(job, instance string) (metadataCache, error) {
+	targetGroupMap := t.sm.TargetsAll()
+	targetGroup, ok := targetGroupMap[job]
+
+	if !ok {
+		//when the job is replaced in relabel_config, TargetsAll() still return the map with old job name as key
+		//so we need to go over all the targets to find the matching job name
+		targetGroup = nil
+	checkJobLoop:
+		for _, potentialTargetGroup := range targetGroupMap {
+			for _, target := range potentialTargetGroup {
+				if target.Labels().Get(model.JobLabel) == job {
+					targetGroup = potentialTargetGroup
+					break checkJobLoop
+				}
+			}
+		}
+
+		if targetGroup == nil {
+			return nil, errors.New("unable to find a target group with job=" + job)
+		}
+	}
+
+	// from the same targetGroup, instance is not going to be duplicated
+	for _, target := range targetGroup {
+		if target.Labels().Get(model.InstanceLabel) == instance {
+			return &mCache{target}, nil
+		}
+	}
+	return nil, errors.New("unable to find a target with job=" + job + ", and instance=" + instance)
+}
+
+type ScrapeManager interface {
+	TargetsAll() map[string][]*scrape.Target
+}
+
+type metricsTypeHandler struct {
+	mc *metadataServiceImpl
```

Comment:
Why not declare it as interface "*metadataService"?

Reply:
Will change it